### ISSUE TITLE
Unable to run Buck2 on Illumos

### ISSUE THREAD

**Issue by davidbarsky** [facebook/buck2#187]
I was chatting with @steveklabnik yesterday, and he reported that after running `RUST_BACKTRACE=full buck2 targets //...` (the only previous command was `buck init --git`) on a Illumos machine, he got the following panic:

<details>

<summary>Backtrace from Buck</summary>

```
sklabnik@atrium ~/omicron $ RUST_BACKTRACE=full buck2 targets //...
thread 'buck2-rt' panicked at 'not implemented: no support yet for operating system `illumos`', app/buck2_server/src/host_info.rs:31:18
stack backtrace:
   0:          0x6930d26 - std::backtrace_rs::backtrace::libunwind::trace::hcacd93644e5ef47f
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:          0x6930d26 - std::backtrace_rs::backtrace::trace_unsynchronized::h9db9225d661ba70a
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:          0x6930d26 - std::sys_common::backtrace::_print_fmt::h88bdd2fdb0131e3b
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/sys_common/backtrace.rs:65:5
   3:          0x6930d26 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h72ec66cd6c137260
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/sys_common/backtrace.rs:44:22
   4:          0x698e74a - core::fmt::write::h8f4cd8a9da8313ae
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/core/src/fmt/mod.rs:1213:17
   5:          0x6922164 - std::io::Write::write_fmt::h27695cc6d6a15888
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/io/mod.rs:1682:15
   6:          0x6930b00 - std::sys_common::backtrace::_print::h235565f55ac31745
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/sys_common/backtrace.rs:47:5
   7:          0x6930b00 - std::sys_common::backtrace::print::hbff4f86db8864c6e
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/sys_common/backtrace.rs:34:9
   8:          0x6933803 - std::panicking::default_hook::{{closure}}::h5e7acbf2a5815ad0
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/panicking.rs:267:22
   9:          0x693349d - std::panicking::default_hook::h6e8b4110d329b5a6
                               at /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/std/src/panicking.rs:286:9
  10:          0x650c9a1 - gazebo::terminate_on_panic::{{closure}}::h65a9d0f267c10361
Build ID: 07225076-624b-4c2e-bb66-d92032656cc8
Jobs completed: 0. Time elapsed: 0.1s.
Command failed: Buck daemon event bus encountered an error, the root cause (if available) is displayed above this message.

Caused by:
    0: status: Unknown, message: "error reading a body from connection: stream closed because of a broken pipe", details: [], metadata: MetadataMap { headers: {} }
    1: error reading a body from connection: stream closed because of a broken pipe
    2: stream closed because of a broken pipe

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   1: futures_util::stream::try_stream::TryStreamExt::try_poll_next_unpin
   2: <futures_util::stream::unfold::Unfold<T,F,Fut> as futures_core::stream::Stream>::poll_next
   3: futures_util::stream::stream::StreamExt::poll_next_unpin
   4: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
   5: <buck2_client::commands::targets::TargetsCommand as buck2_client_ctx::streaming::StreamingCommand>::exec_impl::{{closure}}
   6: futures_util::future::future::FutureExt::poll_unpin
   7: <core::pin::Pin<P> as core::future::future::Future>::poll
   8: tokio::runtime::scheduler::current_thread::Context::enter
   9: tokio::macros::scoped_tls::ScopedKey<T>::set
  10: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
  11: tokio::runtime::runtime::Runtime::block_on
```

</details>

It seems like [`get_host_info`]( https://github.com/facebook/buck2/blob/0ac9355dd53548742c21da13f4d6ccf497b315a1/app/buck2_server/src/host_info.rs#L31) is restrictive with what operating systems are supported. Naively, it seems like a good defensive programming approach, but with some recent changes made to Buck to get it _building_ on Illumos, I suspect that this is a check that can be weakened a little bit. My cursory searching of Buck indicates that the host information is used decently heavily inside of [`BuildInterpreterConfiguror`](https://github.com/facebook/buck2/blob/4da29d3e2f35eaac3a5724a929e348662644df4a/app/buck2_interpreter_for_build/src/interpreter/configuror.rs#L79)—I'm guessing that the fix here might entail adding another enum variant to [`InterpreterHostPlatform`](https://github.com/facebook/buck2/blob/4da29d3e2f35eaac3a5724a929e348662644df4a/app/buck2_interpreter/src/extra/mod.rs#L19).

(I'm writing this up because I was nerd-sniped yesterday and I figured I'd I share what I found so far.)

**Comment by ndmitchell** [comment]
The main use of `get_host_info` is to support the user supplied `host_info()` function (which we'd really rather people used less). I'll put up a diff to turn these into having an `Unknown` variant and we can see what people think.

**Comment by ndmitchell** [comment]
Buck1 already supports host_info.is_unknown, so I've plumbed that through internally and put up a diff. Once it gets approved it should ship out.

**Comment by steveklabnik** [comment]
Amazing, thanks! Always awesome when you go to start figuring out a fix and someone has beat you to it :)

**Comment by ndmitchell** [comment]
Should be fixed since ee741f61b45bd370b0bff7b7dcfb21c26449db5b. Let me know if there are still any issues.

**Comment by steveklabnik** [comment]
```text
$ buck2 targets //...
Build ID: 8d39cc18-b794-4c63-9433-b352995ab31d
Jobs completed: 4. Time elapsed: 0.2s.
root//:hello_world
```

🎊🎊🎊
